Editor: Add highlight around docks when dragging

lodetrick opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/12489

Adds a highlight around docks when dragging that tells you whether you can drop a dock. Also increases the area to drop docks to the entire TabContainer. 

https://github.com/user-attachments/assets/074e7e41-32e1-4b0b-baa2-35c233f834e3

This PR also creates the EditorDockDragHint node to contain the logic.

lodetrick committed:
Add dock dragging area and highlight

Calinou reviewed (commented):
Tested locally, it works as expected. This is a nice usability improvement :slightly_smiling_face: 

My only concern is that the width of the drag border hint is a bit too wide by default. It also varies depending on the editor spacing preset chosen, and it'll be invisible with the Compact spacing preset:

### Default

![image](https://github.com/user-attachments/assets/4c1ec11b-5251-49b4-835f-cc178983e94e)

### Spacious

![image](https://github.com/user-attachments/assets/6161e5cc-1cb6-43d6-a410-4b22f1601436)

### Compact

![image](https://github.com/user-attachments/assets/3fed99a1-b723-4730-8aed-fa3a14c875dc)

I suggest using a fixed value that's identical to the standard focus outline (`Math::round(2 * EDSCALE)`).

Calinou commented on editor/editor_dock_manager.cpp:
```diff
@@ -52,6 +53,115 @@ enum class TabStyle {
+void EditorDockTabContainer::_notification(int p_what) {
+	switch (p_what) {
+		case EditorSettings::NOTIFICATION_EDITOR_SETTINGS_CHANGED: {
+			int _corner_radius = EDSCALE * EDITOR_GET("interface/theme/corner_radius").operator int();
+			if (_corner_radius != dock_drop_highlight->get_corner_radius(CORNER_TOP_LEFT)) {
```
Remove leading underscore for variable name:

```suggestion
			int corner_radius = EDSCALE * EDITOR_GET("interface/theme/corner_radius").operator int();
			if (corner_radius != dock_drop_highlight->get_corner_radius(CORNER_TOP_LEFT)) {
```

Calinou commented on editor/editor_dock_manager.cpp:
```diff
@@ -52,6 +53,115 @@ enum class TabStyle {
+		case NOTIFICATION_THEME_CHANGED: {
+			valid_drop_color = get_theme_color(SNAME("accent_color"), EditorStringName(Editor));
+			invalid_drop_color = get_theme_color(SNAME("error_color"), EditorStringName(Editor));
+			int _dock_margin = get_theme_constant(SNAME("base_margin"), EditorStringName(Editor));
+
+			if (dock_margin != _dock_margin) {
+				dock_margin = _dock_margin;
```
Use a more explicit variable name:

```suggestion
			int new_dock_margin = get_theme_constant(SNAME("base_margin"), EditorStringName(Editor));

			if (dock_margin != new_dock_margin) {
				dock_margin = new_dock_margin;
```

lodetrick commented:
Thanks for the tips! I changed the width to a constant, and removed the `dock_margin` variable because it was only used to keep track of the border width. I draw the highlight outside of the rect for it to not be obscured by the TabBar (I can't change the z-index of the TabBar because the tabbar background would draw over them), so it is slightly smaller on the edge of the window.

Calinou reviewed (approved):
Looks great now!

![image](https://github.com/user-attachments/assets/1402590c-cbd8-4f94-b014-567249b04b67)

KoBeWi commented on editor/editor_dock_manager.cpp:
```diff
@@ -52,6 +53,120 @@ enum class TabStyle {
+			}
+		} break;
+
+		case NOTIFICATION_POSTINITIALIZE:
```
```suggestion
```

KoBeWi commented on editor/editor_dock_manager.cpp:
```diff
@@ -52,6 +53,120 @@ enum class TabStyle {
+
+void EditorDockTabContainer::_notification(int p_what) {
+	switch (p_what) {
+		case EditorSettings::NOTIFICATION_EDITOR_SETTINGS_CHANGED: {
```
Use `check_changed_settings_in_group()`.

KoBeWi commented on editor/editor_dock_manager.cpp:
```diff
@@ -52,6 +53,120 @@ enum class TabStyle {
 
+// Implemented in input(..) as nearly every single child consumes the mouse input.
+void EditorDockTabContainer::input(const Ref<InputEvent> &p_event) {
+	ERR_FAIL_COND(p_event.is_null());
```
Useless check.

KoBeWi commented on editor/editor_dock_manager.cpp:
```diff
@@ -52,6 +53,120 @@ enum class TabStyle {
+	ERR_FAIL_COND(p_event.is_null());
+
+	Ref<InputEventMouse> me = p_event;
+	if (me.is_valid() && is_dragging_dock) {
```
`is_dragging_dock` is always true here, no?

KoBeWi commented on editor/editor_dock_manager.cpp:
```diff
@@ -52,6 +53,120 @@ enum class TabStyle {
+			dock_drop_data = get_viewport()->gui_get_drag_data();
+
+			// Check if we are dragging a dock.
+			if (dock_drop_data.has("type") && dock_drop_data["type"] == "tab_container_tab") {
```
```suggestion
			const String type = dock_drop_data.get("type", "");
			if (type == "tab_container_tab") {
```

KoBeWi commented on editor/editor_dock_manager.cpp:
```diff
@@ -52,6 +53,120 @@ enum class TabStyle {
+
+		// Ideally this would be done directly after gui_input but this node doesn't always receive
+		// gui_input and the input is accepted when dragging, so this can't be done in the input system.
+		case NOTIFICATION_PROCESS: {
```
But why is this not inside `input()`, where `mouse_inside` is set?

lodetrick replied:
Because the mouse cursor is set inside `gui_input()` (specifically `viewport::_gui_input_event()`), and `input()` is called before that, so any changes to the cursor shape will be overridden by the viewport. The only reason to have this is to have the proper cursor shape.

KoBeWi replied:
You could use `call_deferred()` for that.

lodetrick replied:
I'll give that a try

KoBeWi commented on editor/editor_dock_manager.h:
```diff
@@ -97,7 +124,7 @@ class EditorDockManager : public Object {
 
 	Vector<WindowWrapper *> dock_windows;
-	TabContainer *dock_slot[DOCK_SLOT_MAX];
+	EditorDockTabContainer *dock_slot[DOCK_SLOT_MAX];
```
https://github.com/godotengine/godot/pull/106762/files#r2110295167

EditorDockTabContainer's logic is entirely internal, so you can use TabContainer.

kitbdev commented:
The `drop_data` and other methods are intended to only be called from the Viewport, this is fighting with the drag and drop system.
I was imagining dynamically creating new Controls that cover the TabContainer's contents on drag start, this way we can have more drop targets than just in the TabContainers (eg. to move to dock slots with no tabs), the contents wouldn't receive their own can_drop, and it would work with the drag and drop system better. 
Though I'm not sure if there are any problems with that approach, and I wanted to get a better idea of how floating docks could support it before trying it.

lodetrick commented:
Thats a good idea, I'll play around with it. I think that floating docks right now are more about docks themselves rather than the TabContainers

lodetrick commented on editor/editor_dock_manager.cpp:
```diff
@@ -114,6 +197,192 @@ void DockSplitContainer::remove_child_notify(Node *p_child) {
+		int dock_hidden_size = 28 * EDSCALE;
+		int dock_hidden_offset = dock_hidden_size + 4 * EDSCALE;
+		Rect2 temp_dock_drag_rects[DOCK_SLOT_MAX];
+		temp_dock_drag_rects[DOCK_SLOT_LEFT_UL] = dock_slot[DOCK_SLOT_LEFT_UL]->get_global_rect();
```
Unfortunately since `EditorDockManager` is an Object, it does not receive the drag notification. However, I already have safeguards against computing the information multiple times, it stores the output and has an early return if the output is already known for the current drag.

lodetrick replied:
Actually thinking about it, I do think that we should override the TabBar drag and dropping. If we want to limit where the dock can be dropped (ie for different horizontal / vertical layouts), we need to change the behavior somewhat, and I think that this is the place. I might try and see if I can call the TabBar's gui_input manually for updating the tabbar's drop hint.

lodetrick commented:
It should be back to the original design from a user standpoint, but with a control as the backend, cleaning up the implementation a lot. There's a few changes that should be made once the EditorDock PR is merged, but until then it is good.

kitbdev commented on editor/editor_dock_manager.cpp:
```diff
@@ -52,6 +53,86 @@ enum class TabStyle {
+
+	set_as_top_level(true);
+	dock_drop_highlight.instantiate();
+	dock_drop_highlight->set_draw_center(false);
```
Maybe this should draw a partially transparent center so it is more clear that it isn't for the dock's content?

lodetrick commented:
I've cleaned up the code some more, and made the center a transparent color corresponding to the edge color, its subtle but I like it. I also extracted the tab drop drawing logic into its own function so both the TabBar and DragHint classes can use it.

<img width="294" alt="Screenshot 2025-06-02 at 4 21 22 PM" src="https://github.com/user-attachments/assets/e77f238c-b240-49c9-b450-a18c31bb5f44" />

kitbdev commented on editor/editor_dock_manager.cpp:
```diff
@@ -52,6 +53,118 @@ enum class TabStyle {
+			}
+
+			// TODO: Update logic when GH-106503 is merged to use flags.
+			can_drop_dock = is_layout_horizontal ? bool(dragged_dock->call("_can_dock_horizontal")) : true;
```
is_layout_horizontal is always false, so this check doesn't do anything and invalid color is never used, even after #106503. 
We can just set can_drop_dock to true, and remove the invalid color for now. When a PR adds horizontal slots it can be re-added then, if it is wanted.

kitbdev reviewed (approved):
Works as expected. 
Can be improved in the future with dragging to closed slots, and `_get_dock_tab_dragged` logic can be simplified with some kind of Dock manager control.

KoBeWi commented on editor/editor_dock_manager.cpp:
```diff
@@ -52,6 +53,115 @@ enum class TabStyle {
+
+		case NOTIFICATION_DRAW: {
+			// Draw highlights around docks that can be dropped.
+			if (mouse_inside) {
```
Can be early return.

akien-mga commented:
Thanks!
